llama.cpp

```diff
@@ -1011,6 +1016,18 @@ static void llama_model_load_internal(
     ml->load_all_data(progress_callback, progress_callback_user_data, use_mlock ? &lctx.model.mlock_mmap : NULL);
 
     model.mapping = std::move(ml->mapping);
+#ifdef GGML_USE_CUBLAS
+    for (int i = 0; i < std::min(gpu_layers, int(hparams.n_layer)); ++i) {
+        auto & layer = model.layers[i];
+        ggml_cuda_transform_tensor(layer.wq);
+        ggml_cuda_transform_tensor(layer.wk);
+        ggml_cuda_transform_tensor(layer.wv);
+        ggml_cuda_transform_tensor(layer.wo);
+        ggml_cuda_transform_tensor(layer.w1);
+        ggml_cuda_transform_tensor(layer.w2);
+        ggml_cuda_transform_tensor(layer.w3);
+    }
+#endif
```

Comment:
Would it be possible to to have this be a generic API from ggml instead of something tied to CUDA?

Reply:
I see your point, CL would additionally need an offset, maybe.

Reply:
How about this: the `data` property of `ggml_tensor` is a void pointer. For regular tensors and the transformed tensors in this PR this pointer simply points at the data. You could instead point it at whatever OpenCL has for GPU memory pointers and then dereference that pointer in turn. That would cost you one memory access per weight but I think that overhead is negligible.

Reply:
I've been working on this same problem this week, albeit from a different angle.  I've been experimenting with compiling llama.cpp for lower-end devices with integrated GPUs, such as Nvidia Tegra.  On those devices, the memory is physically shared between the CPU and the GPU, allowing us to use a unified address space and leveraging zero-copy when doing GPU processing.

I see a lot of similarities between your approach and things that I've been experimenting with.  One issue that I'm running into is that the `ggml_cuda_h2d_tensor_2d` function does a bit of transformation on the data in addition to copying it over to the GPU, so separating that functionality has made things a bit more difficult for me to accelerate with the zero-copy case.

Reply:
If tensors are kept on device, will that significantly hurt performance when the CPU operates on the graph?

Reply:
This is the KV cache, it is not read by the CPU, but currently the CPU writes to it, when a batch of embeddings is processed on per layer.

Reply:
> Now I'm thinking it would be better to copy the whole tensor or even keep it on device and use something like GemmStridedBatched on it (because this tensor also has the third dimension as well)

If we did this, then it would certainly make the zero-copy case much easier, because we would no longer be transforming and copying in the same motion.